compiler/src/org.graalvm.compiler.nodes/src/org/graalvm/compiler/nodes/calc/IntegerDivRemNode.java

```diff
@@ -83,6 +85,92 @@ public final Type getType() {
         return type;
     }
 
+    protected static ValueNode canonicalizeDivConstant(ValueNode forX, long c, NodeView view) {
+        boolean isDivisorPositive = c >= 0;
+        long divisor = isDivisorPositive ? c : -c;
+        int bitSize = forX.stamp(view).getStackKind().getBitCount();
+        Pair<Long, Integer> nums = magicDivideConstants(divisor, bitSize);
+        if (nums == null) {
+            return null;
+        }
+
+        long magicNum = nums.getLeft().longValue();
+        int shiftNum = nums.getRight().intValue();
+        ValueNode value;
+        if (bitSize == Integer.SIZE) {
+            value = new MulNode(new SignExtendNode(forX, Long.SIZE), ConstantNode.forLong(magicNum));
+            if (magicNum < 0) {
+                value = new RightShiftNode(value, ConstantNode.forInt(bitSize));
+                value = new AddNode(forX, new NarrowNode(value, bitSize));
+                if (shiftNum != 0) {
+                    value = new RightShiftNode(value, ConstantNode.forInt(shiftNum));
+                }
+            } else {
+                value = new RightShiftNode(value, ConstantNode.forInt(bitSize + shiftNum));
+                value = new NarrowNode(value, bitSize);
+            }
+        } else {
+            assert bitSize == Long.SIZE;
+            value = new IntegerMulHighNode(forX, ConstantNode.forLong(magicNum));
+            if (magicNum < 0) {
+                value = new AddNode(forX, value);
+            }
+            if (shiftNum != 0) {
+                value = new RightShiftNode(value, ConstantNode.forInt(shiftNum));
+            }
+        }
+
+        ValueNode subValue = new RightShiftNode(forX, ConstantNode.forInt(bitSize - 1));
+        if (isDivisorPositive) {
+            return new SubNode(value, subValue);
+        }
+        return new SubNode(subValue, value);
+    }
+
+    /**
+     * Borrowed from Hacker's Delight by Henry S. Warren, Jr.
+     */
+    private static Pair<Long, Integer> magicDivideConstants(long divisor, int size) {
+        if (divisor == 0 || divisor == 1) {
+            return null;
+        }
+
+        final long twoW = 1L << (size - 1);                // 2 ^ (size - 1).
+        long t = twoW + (divisor >>> (size - 1));
+        long ad = Math.abs(divisor);
```

Comment:
Yes, the algorithm accepts negative divisor. But I think we'd better use the absolute divisor because the generated magic number will be negative under two situations:
1. If the divisor is positive, and `2^(w-1) <= M < 2^w`, then the magic number is` M - 2^w`.
2. If the divisor is negative, and `0 <= M < 2^(w-1)`, then the magic number is `-M`.
For the first situation, we need to insert some instructions to correct the final result, like the following codes does:
`if (magicNum < 0) {
    value = new RightShiftNode(value, ConstantNode.forInt(bitSize));
    value = new AddNode(forX, new NarrowNode(value, bitSize));
    if (shiftNum != 0) {
        value = new RightShiftNode(value, ConstantNode.forInt(shiftNum));
    }
} else {
     value = new RightShiftNode(value, ConstantNode.forInt(bitSize + shiftNum));
     value = new NarrowNode(value, bitSize);
}`

So to distinguish the situations, I think one easy way is to keep the divisor positive.